### ISSUE TITLE
tf.keras.losses.binary_crossentropy gives worse result than calling 'binary_crossentropy'

### ISSUE THREAD

**Issue by chrismaliszewski** [tensorflow/tensorflow#41878]
**QUESTION 1:** Why does calling binary_crossentropy by its name with a string 'binary_crossentropy' gives different (better) results than calling, the same(?), tf.keras.losses.binary_crossentropy function inside a custom loss function?

**QUESTION 2:** What am I doing wrong?

**PROBLEM:** I'm trying to solve a mystery of different results I get when I use:
```python
def bc_custom(y_true, y_pred): return tf.keras.losses.binary_crossentropy(y_true, y_pred)
model1.compile(optimizer=tf.compat.v2.optimizers.Adam(learning_rate=LEARNING_RATE, loss = bc_custom, metrics=['binary_accuracy'])
```
and 
```python
model2.compile(optimizer=tf.compat.v2.optimizers.Adam(learning_rate=LEARNING_RATE, loss = 'binary_crossentropy', metrics=['binary_accuracy'])
```
Basically: 

 - the difference is in calling binary_crossentropy inside a custom function (model1) and the same function by its name with a string (model2),
 - following https://stackoverflow.com/a/57331394/5536388 advice, I set metrics to binary_accuracy,
 - both models use the same dataset which may only differ after being shuffled in the training process,
 - the models solve a multilabel classification problem. The final layer uses the sigmoid activation function.

**RESULTS:**
I made several tests and each time model1:

 - is being taught much slower:

 epoch, accuracy, loss, accuracy_val, loss_val
    
    0,0.01571397,0.08330947177696847,0.08565965,0.06395021689980937
    1,0.14730678,0.04977951637815227,0.24956022,0.039584608338986936 
    2,0.2366954,0.98736,0.2366954,0.03144164173531004,0.27801147,0.99116695,0.27801147,0.022898457557974424

vs model2

    0,0.98126763,0.08199372373072011,0.9798808,0.061917345579108135
    1,0.98271894,0.0477434622772295,0.98628557,0.03482157470263099
    2,0.9884866,0.9884866,0.24891642,0.028994456771504733,0.9914624,0.9914624,0.27778202,0.022188534344907987

 - model1 gives worse results, at least in the beginning (haven't waited till the final results):
[![model1 results after epoch 0][1]][1]
results of model1 after epoch 0
[![model2 results after epoch 0][2]][2]
results of model2 after epoch 0

 - what's even more interesting, when a smaller dataset, 'binary_crossentropy' still gives some prediction results while the custom function returns 0 probability across all labels :|. 

The expected results are one-point-wide straight-to-top lines in several positions (10, 30, 50, 110, 140, 210, 230) as the model2 is clearly going to and, in the end, reaches the correct answer in 98% of cases. 

  [1]: https://i.stack.imgur.com/llgpH.png
  [2]: https://i.stack.imgur.com/eFg5l.png

**Comment by ravikyram** [comment]
@chrismaliszewski 

Request you to share colab link or simple stand alone code to reproduce the issue in our environment.It helps us in localizing the issue faster.Thanks!

**Comment by chrismaliszewski** [comment]
@ravikyram, which issue template should I use in your opinion? I wasn't sure since I don't think it's a bug nor any other from the list of templates. That's why I created "Other issues". 

It's impossible for me to share the data I have since it's too big (39GB). My input is 256x50x3 x ~120k images. My model is VGG16 without top layers + flatten + 3x dense/dropout (8192, 0.5) + dense (256, sigmoid). 

I use TF 2.1.0 on Windows in Anaconda.

**Comment by chrismaliszewski** [comment]
So here is a piece of code that briefly presents what I do: https://colab.research.google.com/drive/1aTG3ApEUmr9UyhSUoALY-_wEgTTOHlaz?usp=sharing

So:

1. loading data, preparing data (I preprocess my arrays for VGG16), divided it into train-test datasets 90:10 + ~10 extra for checking predictions,
2. Training.
3. (not shown in the code above) Checking predictions with each epoch's end. 

Unfortunately, the problem does not appear in the above code but it does in my case.

**QUESTION:** Should there be any difference between calling 'binary_crossentropy' or tf.keras.losses.binary_crossentropy? What I mean is, does 'binary_crossentropy' calls the function without anything else or is there anything else done on the way?

**Comment by jvishnuvardhan** [comment]
@chrismaliszewski Is this still an issue for you? I ran your code with recent `tf-nightly` and I don't see much difference between two models. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/500f5581afe9a1468a30383d43480d5f/-binary_accuracy-vs-tf-keras-losses-binary_crossentropy.ipynb).

Please note that I reduced training data and didn't consider `validation_data` to minimize computational time. Thanks!

Please verify and close the issue if this was resolved for you. Thanks!

**Comment by chrismaliszewski** [comment]
Hi. No, the issue still exists. I've decided to move on and I'm transferring my code to Pytorch since it is another problem on the list with TF I have and I'm hoping the problems don't exist there. Thank you for your concern. I'm closing the issue.

**Comment by tanzhenyu** [comment]
Do you use sample_weights?

**Comment by chrismaliszewski** [comment]
No, training from scratch in both cases, using a custom loss function which returns the results from tf.keras.losses.binary_crossentropy and 'binary_crossentropy'.
